src/library/type_context.cpp

```diff
@@ -658,7 +659,7 @@ bool type_context::should_unfold_macro(expr const & e) {
 
 optional<expr> type_context::expand_macro(expr const & e) {
     lean_assert(is_macro(e));
-    if (should_unfold_macro(e)) {
+    if (should_unfold_macro(e) && !is_name_macro(e)) {
```

Comment:
I believe it should be fixed by `whnf_core` also respecting `m_unfold_pred`, if that makes sense.